File: src/batch.js
/**
 * Request Batching Module
 * Combine multiple RPC calls into single HTTP request
 */

import http from 'http';
import https from 'https';
import { getPool } from './pool.js';

class BatchRequest {
  constructor(rpcUrl, options = {}) {
    this.rpcUrl = rpcUrl;
    this.pool = getPool();
    this.options = {
      maxBatchSize: 100,
      timeout: 30000,
      autoFlush: true,
      flushInterval: 50,
      ...options
    };

    this.queue = [];
    this.pending = new Map();
    this.idCounter = 1;
    this.flushTimer = null;
  }

  add(method, params = []) {
    return new Promise((resolve, reject) => {
      const id = this.idCounter++;
      
      this.queue.push({
        jsonrpc: '2.0',
        id,
        method,
        params
      });

      this.pending.set(id, { resolve, reject });

      if (this.queue.length >= this.options.maxBatchSize) {
        this.flush();
      } else if (this.options.autoFlush && !this.flushTimer) {
        this.flushTimer = setTimeout(() => {
          this.flush();
        }, this.options.flushInterval);
      }
    });
  }

  async flush() {
    if (this.flushTimer) {
      clearTimeout(this.flushTimer);
      this.flushTimer = null;
    }

    if (this.queue.length === 0) {
      return { success: true, results: [], count: 0 };
    }

    const batch = [...this.queue];
    const pendingCopy = new Map(this.pending);
    this.queue = [];
    this.pending.clear();

    const startTime = performance.now();

    try {
      const responses = await this._sendBatch(batch);
      const latency = performance.now() - startTime;

      if (Array.isArray(responses)) {
        for (const response of responses) {
          const pending = pendingCopy.get(response.id);
          if (pending) {
            if (response.error) {
              pending.reject(new Error(response.error.message || 'RPC Error'));
            } else {
              pending.resolve(response.result);
            }
            pendingCopy.delete(response.id);
          }
        }
      }

      for (const [id, pending] of pendingCopy) {
        pending.reject(new Error('No response received'));
      }

      return {
        success: true,
        count: batch.length,
        latency: Math.round(latency * 100) / 100
      };
    } catch (error) {
      for (const [id, pending] of pendingCopy) {
        pending.reject(error);
      }

      return {
        success: false,
        count: batch.length,
        error: error.message
      };
    }
  }

  async _sendBatch(batch) {
    const agent = this.pool.getAgent(this.rpcUrl);
    const parsed = new URL(this.rpcUrl);
    const isHttps = parsed.protocol === 'https:';

    const payload = JSON.stringify(batch);

    return new Promise((resolve, reject) => {
      const httpModule = isHttps ? https : http;

      const req = httpModule.request(
        {
          hostname: parsed.hostname,
          port: parsed.port || (isHttps ? 443 : 80),
          path: parsed.pathname,
          method: 'POST',
          agent,
          timeout: this.options.timeout,
          headers: {
            'Content-Type': 'application/json',
            'Content-Length': Buffer.byteLength(payload)
          }
        },
        (res) => {
          let data = '';
          res.on('data', chunk => data += chunk);
          res.on('end', () => {
            try {
              const json = JSON.parse(data);
              resolve(json);
            } catch (e) {
              reject(new Error('Invalid JSON response'));
            }
          });
        }
      );

      req.on('error', reject);
      req.on('timeout', () => {
        req.destroy();
        reject(new Error('Batch request timeout'));
      });

      req.write(payload);
      req.end();
    });
  }

  get size() {
    return this.queue.length;
  }

  clear() {
    if (this.flushTimer) {
      clearTimeout(this.flushTimer);
      this.flushTimer = null;
    }

    for (const [id, pending] of this.pending) {
      pending.reject(new Error('Batch cleared'));
    }

    this.queue = [];
    this.pending.clear();
  }
}

class BatchBuilder {
  constructor(rpcUrl) {
    this.rpcUrl = rpcUrl;
    this.calls = [];
  }

  call(method, params = []) {
    this.calls.push({ method, params });
    return this;
  }

  ethCall(to, data, blockTag = 'latest') {
    return this.call('eth_call', [{ to, data }, blockTag]);
  }

  getBalance(address, blockTag = 'latest') {
    return this.call('eth_getBalance', [address, blockTag]);
  }

  getNonce(address, blockTag = 'latest') {
    return this.call('eth_getTransactionCount', [address, blockTag]);
  }

  blockNumber() {
    return this.call('eth_blockNumber', []);
  }

  gasPrice() {
    return this.call('eth_gasPrice', []);
  }

  getBlock(blockNumber = 'latest', fullTx = false) {
    return this.call('eth_getBlockByNumber', [blockNumber, fullTx]);
  }

  getSlot() {
    return this.call('getSlot', []);
  }

  solanaGetBalance(pubkey) {
    return this.call('getBalance', [pubkey]);
  }

  async execute() {
    if (this.calls.length === 0) {
      return { success: true, results: [], latency: 0 };
    }

    const pool = getPool();
    const agent = pool.getAgent(this.rpcUrl);
    const parsed = new URL(this.rpcUrl);
    const isHttps = parsed.protocol === 'https:';

    const batch = this.calls.map((call, index) => ({
      jsonrpc: '2.0',
      id: index + 1,
      method: call.method,
      params: call.params
    }));

    const payload = JSON.stringify(batch);
    const startTime = performance.now();

    return new Promise((resolve) => {
      const httpModule = isHttps ? https : http;

      const req = httpModule.request(
        {
          hostname: parsed.hostname,
          port: parsed.port || (isHttps ? 443 : 80),
          path: parsed.pathname,
          method: 'POST',
          agent,
          timeout: 30000,
          headers: {
            'Content-Type': 'application/json',
            'Content-Length': Buffer.byteLength(payload)
          }
        },
        (res) => {
          let data = '';
          res.on('data', chunk => data += chunk);
          res.on('end', () => {
            const latency = performance.now() - startTime;
            try {
              const json = JSON.parse(data);
              
              const sorted = Array.isArray(json) 
                ? json.sort((a, b) => a.id - b.id)
                : [json];

              const results = sorted.map(r => ({
                success: !r.error,
                result: r.result,
                error: r.error?.message
              }));

              resolve({
                success: true,
                results,
                count: this.calls.length,
                latency: Math.round(latency * 100) / 100
              });
            } catch (e) {
              resolve({
                success: false,
                error: 'Invalid JSON response',
                count: this.calls.length,
                latency: Math.round(latency * 100) / 100
              });
            }
          });
        }
      );

      req.on('error', (err) => {
        resolve({
          success: false,
          error: err.message,
          count: this.calls.length,
          latency: Math.round((performance.now() - startTime) * 100) / 100
        });
      });

      req.on('timeout', () => {
        req.destroy();
        resolve({
          success: false,
          error: 'Timeout',
          count: this.calls.length,
          latency: 30000
        });
      });

      req.write(payload);
      req.end();
    });
  }
}

export function createBatchRequest(rpcUrl, options) {
  return new BatchRequest(rpcUrl, options);
}

export function batch(rpcUrl) {
  return new BatchBuilder(rpcUrl);
}

export default BatchRequest;

File: rust/src/batch.rs
//! Request Batching Module
//!
//! Combine multiple RPC calls into single HTTP request

use anyhow::Result;
use serde::{Deserialize, Serialize};
use std::time::{Duration, Instant};

#[derive(Debug, Clone, Serialize)]
pub struct BatchRpcRequest {
    pub jsonrpc: String,
    pub id: u64,
    pub method: String,
    pub params: Vec<serde_json::Value>,
}

#[derive(Debug, Clone, Deserialize)]
pub struct BatchRpcResponse {
    pub jsonrpc: String,
    pub id: u64,
    pub result: Option<serde_json::Value>,
    pub error: Option<BatchRpcError>,
}

#[derive(Debug, Clone, Deserialize)]
pub struct BatchRpcError {
    pub code: i64,
    pub message: String,
}

#[derive(Debug, Clone)]
pub struct BatchResult {
    pub success: bool,
    pub result: Option<serde_json::Value>,
    pub error: Option<String>,
}

#[derive(Debug, Clone)]
pub struct BatchExecuteResult {
    pub success: bool,
    pub results: Vec<BatchResult>,
    pub count: usize,
    pub latency_ms: f64,
    pub error: Option<String>,
}

/// Batch request builder
pub struct BatchBuilder {
    rpc_url: String,
    calls: Vec<(String, Vec<serde_json::Value>)>,
    client: reqwest::Client,
    timeout_ms: u64,
}

impl BatchBuilder {
    /// Create new batch builder
    pub fn new(rpc_url: &str) -> Self {
        let client = reqwest::Client::builder()
            .timeout(Duration::from_secs(30))
            .pool_max_idle_per_host(10)
            .build()
            .expect("Failed to create HTTP client");

        Self {
            rpc_url: rpc_url.to_string(),
            calls: Vec::new(),
            client,
            timeout_ms: 30000,
        }
    }

    /// Set custom timeout
    pub fn timeout(mut self, timeout_ms: u64) -> Self {
        self.timeout_ms = timeout_ms;
        self
    }

    /// Add generic RPC call
    pub fn call(mut self, method: &str, params: Vec<serde_json::Value>) -> Self {
        self.calls.push((method.to_string(), params));
        self
    }

    /// Add eth_call
    pub fn eth_call(self, to: &str, data: &str, block_tag: &str) -> Self {
        let params = vec![
            serde_json::json!({ "to": to, "data": data }),
            serde_json::json!(block_tag),
        ];
        self.call("eth_call", params)
    }

    /// Add eth_getBalance
    pub fn get_balance(self, address: &str, block_tag: &str) -> Self {
        let params = vec![
            serde_json::json!(address),
            serde_json::json!(block_tag),
        ];
        self.call("eth_getBalance", params)
    }

    /// Add eth_getTransactionCount (nonce)
    pub fn get_nonce(self, address: &str, block_tag: &str) -> Self {
        let params = vec![
            serde_json::json!(address),
            serde_json::json!(block_tag),
        ];
        self.call("eth_getTransactionCount", params)
    }

    /// Add eth_blockNumber
    pub fn block_number(self) -> Self {
        self.call("eth_blockNumber", vec![])
    }

    /// Add eth_gasPrice
    pub fn gas_price(self) -> Self {
        self.call("eth_gasPrice", vec![])
    }

    /// Add eth_getBlockByNumber
    pub fn get_block(self, block_number: &str, full_tx: bool) -> Self {
        let params = vec![
            serde_json::json!(block_number),
            serde_json::json!(full_tx),
        ];
        self.call("eth_getBlockByNumber", params)
    }

    /// Add eth_chainId
    pub fn chain_id(self) -> Self {
        self.call("eth_chainId", vec![])
    }

    /// Add eth_estimateGas
    pub fn estimate_gas(self, to: &str, data: &str, value: Option<&str>) -> Self {
        let mut tx = serde_json::json!({ "to": to, "data": data });
        if let Some(v) = value {
            tx["value"] = serde_json::json!(v);
        }
        self.call("eth_estimateGas", vec![tx])
    }

    /// Add Solana getSlot
    pub fn get_slot(self) -> Self {
        self.call("getSlot", vec![])
    }

    /// Add Solana getBalance
    pub fn solana_get_balance(self, pubkey: &str) -> Self {
        let params = vec![serde_json::json!(pubkey)];
        self.call("getBalance", params)
    }

    /// Add Solana getLatestBlockhash
    pub fn get_latest_blockhash(self) -> Self {
        let params = vec![serde_json::json!({ "commitment": "finalized" })];
        self.call("getLatestBlockhash", params)
    }

    /// Get number of calls in batch
    pub fn len(&self) -> usize {
        self.calls.len()
    }

    /// Check if batch is empty
    pub fn is_empty(&self) -> bool {
        self.calls.is_empty()
    }

    /// Execute batch request
    pub async fn execute(self) -> BatchExecuteResult {
        if self.calls.is_empty() {
            return BatchExecuteResult {
                success: true,
                results: vec![],
                count: 0,
                latency_ms: 0.0,
                error: None,
            };
        }

        // Build batch payload
        let batch: Vec<BatchRpcRequest> = self
            .calls
            .iter()
            .enumerate()
            .map(|(i, (method, params))| BatchRpcRequest {
                jsonrpc: "2.0".to_string(),
                id: (i + 1) as u64,
                method: method.clone(),
                params: params.clone(),
            })
            .collect();

        let count = batch.len();
        let start = Instant::now();

        // Send request
        let response = match self.client.post(&self.rpc_url).json(&batch).send().await {
            Ok(r) => r,
            Err(e) => {
                return BatchExecuteResult {
                    success: false,
                    results: vec![],
                    count,
                    latency_ms: start.elapsed().as_secs_f64() * 1000.0,
                    error: Some(e.to_string()),
                };
            }
        };

        // Parse response
        let responses: Vec<BatchRpcResponse> = match response.json().await {
            Ok(r) => r,
            Err(e) => {
                return BatchExecuteResult {
                    success: false,
                    results: vec![],
                    count,
                    latency_ms: start.elapsed().as_secs_f64() * 1000.0,
                    error: Some(format!("Invalid JSON response: {}", e)),
                };
            }
        };

        let latency_ms = start.elapsed().as_secs_f64() * 1000.0;

        // Sort by id and map results
        let mut sorted_responses = responses;
        sorted_responses.sort_by_key(|r| r.id);

        let results: Vec<BatchResult> = sorted_responses
            .into_iter()
            .map(|r| {
                if let Some(error) = r.error {
                    BatchResult {
                        success: false,
                        result: None,
                        error: Some(error.message),
                    }
                } else {
                    BatchResult {
                        success: true,
                        result: r.result,
                        error: None,
                    }
                }
            })
            .collect();

        BatchExecuteResult {
            success: true,
            results,
            count,
            latency_ms: round_2(latency_ms),
            error: None,
        }
    }
}

/// Create new batch builder
pub fn batch(rpc_url: &str) -> BatchBuilder {
    BatchBuilder::new(rpc_url)
}

fn round_2(val: f64) -> f64 {
    (val * 100.0).round() / 100.0
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_batch_builder() {
        let b = batch("https://eth.llamarpc.com")
            .block_number()
            .gas_price()
            .get_balance("0x0000000000000000000000000000000000000000", "latest");

        assert_eq!(b.len(), 3);
    }
}